core/src/main/java/org/infinispan/notifications/Listenable.java

```diff
@@ -17,20 +21,40 @@
     * See the {@link Listener} annotation for more information.
     * <p/>
     *
-    * @param listener must not be null.
+    * @param listener listener to add, must not be null
+    */
+   default void addListener(Object listener) {
+      CompletionStages.join(addListenerAsync(listener));
+   }
+
+   /**
+    * Asynchronous version of {@link #addListener(Object)}
+    * @param listener listener to add, must not be null
+    * @return CompletionStage that when complete the listener is fully installed
     */
-   void addListener(Object listener);
+   CompletionStage<Void> addListenerAsync(Object listener);
 
    /**
     * Removes a listener from the component.
     *
     * @param listener listener to remove.  Must not be null.
     * @throws org.infinispan.IllegalLifecycleStateException may be thrown if the {@code Listenable} is stopped.
     */
-   void removeListener(Object listener);
+   default void removeListener(Object listener) {
+      CompletionStages.join(removeListenerAsync(listener));
+   }
+
+   /**
+    * Asynchronous version of {@link #removeListener(Object)}
+    * @param listener listener to remove, must not be null
+    * @return CompletionStage that when complete the listener is fully removed
+    */
+   CompletionStage<Void> removeListenerAsync(Object listener);
 
    /**
     * @return a set of all listeners registered on this component.
     */
    Set<Object> getListeners();
+
+   boolean hasListener(Class<? extends Annotation> annotationClass);
```

Comment:
Do users need this?

Reply:
Hrmm, probably not I can move it up the inheritance tree.